x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticKnnVectorQueryRewriteInterceptor.java

```diff
@@ -0,0 +1,161 @@
+        QueryBuilder queryBuilder,
+        InferenceIndexInformationForField indexInformation
+    ) {
+        assert (queryBuilder instanceof KnnVectorQueryBuilder);
+        KnnVectorQueryBuilder knnVectorQueryBuilder = (KnnVectorQueryBuilder) queryBuilder;
+        Map<String, List<String>> inferenceIdsIndices = indexInformation.getInferenceIdsIndices();
+
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        boolQueryBuilder.should(
+            createSubQueryForIndices(
+                indexInformation.nonInferenceIndices(),
+                createSubQueryForIndices(indexInformation.nonInferenceIndices(), knnVectorQueryBuilder)
+            )
+        );
+        // We always perform nested subqueries on semantic_text fields, to support
+        // knn queries using query vectors.
+        for (String inferenceId : inferenceIdsIndices.keySet()) {
+            boolQueryBuilder.should(
+                createSubQueryForIndices(
+                    inferenceIdsIndices.get(inferenceId),
+                    buildNestedQueryFromKnnVectorQuery(knnVectorQueryBuilder, inferenceId)
+                )
+            );
+        }
+        return boolQueryBuilder;
+    }
+
+    private QueryBuilder buildNestedQueryFromKnnVectorQuery(QueryBuilder queryBuilder, String searchInferenceId) {
+        assert (queryBuilder instanceof KnnVectorQueryBuilder);
+        KnnVectorQueryBuilder knnVectorQueryBuilder = (KnnVectorQueryBuilder) queryBuilder;
+        TextEmbeddingQueryVectorBuilder queryVectorBuilder = getTextEmbeddingQueryBuilderFromQuery(queryBuilder);
+        if (queryVectorBuilder != null && queryVectorBuilder.getModelId() == null && searchInferenceId != null) {
+            // If the model ID was not specified, we infer the inference ID associated with the semantic_text field.
+            queryVectorBuilder = new TextEmbeddingQueryVectorBuilder(searchInferenceId, queryVectorBuilder.getModelText());
+        }
+        return QueryBuilders.nestedQuery(
+            SemanticTextField.getChunksFieldName(knnVectorQueryBuilder.getFieldName()),
+            buildNewKnnVectorQuery(
+                SemanticTextField.getEmbeddingsFieldName(knnVectorQueryBuilder.getFieldName()),
+                knnVectorQueryBuilder,
+                queryVectorBuilder
+            ),
+            ScoreMode.Max
+        );
+    }
+
+    private TextEmbeddingQueryVectorBuilder getTextEmbeddingQueryBuilderFromQuery(QueryBuilder queryBuilder) {
+        assert (queryBuilder instanceof KnnVectorQueryBuilder);
+        KnnVectorQueryBuilder knnVectorQueryBuilder = (KnnVectorQueryBuilder) queryBuilder;
+        QueryVectorBuilder queryVectorBuilder = knnVectorQueryBuilder.queryVectorBuilder();
+        if (queryVectorBuilder == null) {
+            return null;
+        }
+        assert (queryVectorBuilder instanceof TextEmbeddingQueryVectorBuilder);
+        return (TextEmbeddingQueryVectorBuilder) knnVectorQueryBuilder.queryVectorBuilder();
```

Comment:
I think we can clean this up instead of calling `knnVectorQueryBuilder.queryVectorBuilder()` twice